### PR TITLE
[2.0.0] Transports

### PR DESCRIPTION
**Pull request by andrewmclagan**
### Description

* Abstracting transports layer to allow end-user developers to implement their own. 
* New test harness with extended tests and coverage
* Following `Request`, `Response` javascript standard APIs (see https://developer.mozilla.org/en-US/docs/Web/API/Request) 
* Reduce bundle size by removing unneeded dependancies.
* Packages published under `@yllet/package-name` npm scope.

See https://github.com/ylletjs/yllet/issues/28 for discussion details

### CODE REVIEW

**Comment by andrewmclagan** [comment]
@frozzare Just creating this PR as a holding place for further discussions moving forward.

Currently going to push this for a `v2.0.0-alpha.1` release. Thoughts?

**Comment by frozzare** [comment]
Looks good! Creating a alpha version is great, then we really can test it and not rushing a new version.

1. Do not mix single and double quotes, just use single quotes 👍 
2. Why Drone instead of GitHub Actions (or travis)? (I never used Drone)
3. I don't like the naming of the core package, better to use `Client` since that's what is exported, otherwise I think we shouldn't export is as a default export. But I like `Client` better.

```
import { Client } from '@yllet/core';
import Client from '@yllet/client';
```

**Comment by andrewmclagan** [comment]
Great :-) having allot of of with this project. You have done great work here.

1. Sure that sounds great, shall we include a prettier `.prettierrc` file? That way we can have automatic formatting on commit / version. So many devs have different preferences with formatting. I will be guided by you of course. - Personally I use `cms + shift + f` in VSC that auto formats code using prettier.

2. Drone is a pure Docker CI/CD platform that has huge traction at the moment, spinning up a pipeline is super super easy. Have a look at how much faster that pipeline runs, its allot faster, and its also testing twice the amount of node platforms as well. Basically IMO its easier to read and understand them Travis, it follows the new Cloud Native Foundation specs for CI pipeline definitions. Also if you look around at major libraries, non of them use Travis anymore as its really aged: *React* migrated from Travis to CircleCI a couple years ago, Babel migrated from travis to CircleCI about the same time, MaterialUI migrated to Azure Pipelines about 3 months ago. All these CI use the same YAML definitions, are docker based and can be migrated to and from easily. Im happy to go with Travis, although it could be a great opportunity to learn a modern CI tool? (Its amazing trust me :)). Using a docker based solution will make automated browser testing across versions of IE and such very easy (we should be doing that before alpha goes to beta).

3. 100% agree. core makes no sense. client makes allot more sense:

```javascript
import Client from '@yllet/client';
import FetchTransport from '@yllet/transport-fetch';
```

**Comment by andrewmclagan** [comment]
@frozzare One question. What does Yllet mean? :-) it sounds scandinavian?

**Comment by andrewmclagan** [comment]
See above commit for implementing those suggestions.

**Comment by andrewmclagan** [comment]
Okay, getting very close to an early alpha. I will publish the packages, that way we can start using it in our downstream apps and report any real-world issues / bugs. My team will be able to implement it asap.

**Comment by andrewmclagan** [comment]
@frozzare Interesting: just implemented basic browser tests: IE11 fails. Will resolve it.

**Comment by frozzare** [comment]
2. I'm not sure about Drone.. sure I sounds great but I wouldn't change just because other packages use it. Sure travis may not be top notch but it works. If we just need to run unit tests I see no point to move right now. If we need to run browser test you can do this we Travis (not sure about GitHub actions) we maybe should move, but I will not have the time to look at Drone right now and learn it. (CircleCI config sucks).

1+3 👍 

It's great that you take the library to next level but you change a lot right now, it will take me some time to get into rollup, drone and other stuffs that I have never used. Use more webpack, travis, github actions and so on. 

But it's great that you work on this, since it's not being that maintained the last two years 👍 

Yllet = Ylle = wool. Think I got the name from my waschmaschine.

**Comment by frozzare** [comment]
Added yllet repo to drone so you can test it https://cloud.drone.io/ylletjs/yllet/

**Comment by andrewmclagan** [comment]
Thanks, totally understand that this is ... massive ... changes. Its adding _allot_ to the package. The main reason Im putting so much work into this library is - my team needs it. The current offerings for WP-API SDKs are IMO not great. Not reliable. Not very cross browser.

We were going to write our own.

The only reason I forked this is simple: I love open source, its always better to attempt to work with people. If you are not happy with the changes, we can adjust, although only to a point, we would simple fork and rename if the changes are not going to work for you.

That said - let's keep this dialog open.

**Comment by frozzare** [comment]
The changes are great, but as I said before it's going very fast and I'm not going to have the time to learn everything new right now. But it's better to maintain this than letting die slowly! That's why I added you the org.

So let's keep working on this :) 

Sure Drone looks and will probably use it later on since I use a lot of docker.

**Comment by andrewmclagan** [comment]
Okay, testing suite is complete. Very extensive:

* Unit tests with over 95% coverage (will fail on less)
* Integration tests against a live WP-JSON API (excellent smoke tests)
* Browser tests against `.browserlistrc` supporting down to IE 11 using BrowserStack / Karma
* Full test suite ran against `esm` and `cjs` builds
* Integrated into CI pipeline.

This is about as good as a test suite can get.

File: README.md [hunk 1]
```diff
@@ -1,7 +1,9 @@
 # Yllet
 
+[![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)]
 [![npm](https://img.shields.io/npm/v/yllet.svg)](http://npm.im/yllet)
-[![Build Status](https://travis-ci.org/ylletjs/yllet.svg?branch=master)](https://travis-ci.org/ylletjs/yllet)
+[![Build Status](https://cloud.drone.io/api/badges/beam-australia/yllet/status.svg?ref=refs/heads/transports)](https://cloud.drone.io/beam-australia/yllet)
```
Comment (frozzare) [inline README.md]: Should probably change this to https://cloud.drone.io/ylletjs/yllet

**Comment by andrewmclagan** [comment]
Good catch, the readme needs updating also. I have updated those.

**Comment by andrewmclagan** [comment]
@frozzare How would you feel about granting write permissions to the repo, and drone?

**Comment by andrewmclagan** [comment]
Our team have decided; against my best advice to release it as is for the time being:

https://github.com/wp-fetch/wp-fetch

**Comment by frozzare** [comment]
Sad to here.

I think we still can merge this and I'm sorry that I'm answering so late. Haven't had time to be active on GitHub for the last days.

**Comment by andrewmclagan** [comment]
thats ok, no problem. For sure we can merge it, this is only temporary. When we complete the PR we can start discussing the merge :-)

**Comment by frozzare** [comment]
@andrewmclagan can you fix so I can publish scoped packages in npm? Can’t do anything with the new packages. Also emailed you and asked you about this.

**Comment by frozzare** [comment]
@andrewmclagan ping. Would be really happy if could fix this soon.

**Pull request merged**